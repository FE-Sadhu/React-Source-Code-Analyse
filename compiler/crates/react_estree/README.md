# react_estree

This crate is a Rust representation of the [ESTree format](https://github.com/estree/estree/tree/master) and
popular extenions including JSX and (eventually) Flow and TypeScript.

This crate is intended as the main interchange format with outside code. A typical integration with React Compiler
will look as follows:

1. Host Compiler parses into the host AST format.
2. Host Compiler converts into `react_estree`.
3. Host Compiler invokes React Compiler to compile the input, which (conceptually) 
   returns the resulting code in `react_estree` format.
4. Host Compiler convert back from `react_estree` to its host AST format.

Because React Compiler is intended to support JavaScript-based toolchains, `react_estree` is designed to support
accurate serialization to/from estree-compatible JSON. We may also support the Babel AST format 
(a variant of ESTree) as well, depending on demand.