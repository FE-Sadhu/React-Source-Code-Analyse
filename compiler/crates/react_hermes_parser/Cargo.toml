[package]
name = "react_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
react_diagnostics = { workspace = true }
react_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
react_estree_codegen = { workspace = true }