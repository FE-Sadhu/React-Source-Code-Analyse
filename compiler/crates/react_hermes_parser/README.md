# react_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `react_estree` AST.