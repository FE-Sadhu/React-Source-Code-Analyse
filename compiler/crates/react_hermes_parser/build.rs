/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use react_estree_codegen::estree_hermes;

// Example custom build script.
fn main() {
    // Re-run if the codegen files change
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/codegen.rs");
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/lib.rs");
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/ecmascript.json");
    println!("cargo:rerun-if-changed=../react_estree_codegen");

    let src = estree_hermes();
    let copyright = "
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
    "
    .to_string();
    let trimmed_copyright = copyright.trim();
    let contents = format!("{trimmed_copyright}\n{src}");
    std::fs::write("src/generated.rs", contents).unwrap();
}
