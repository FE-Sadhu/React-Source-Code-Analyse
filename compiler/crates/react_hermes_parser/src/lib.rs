/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

mod generated;
mod generated_extension;

use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;
use react_diagnostics::Diagnostic;
use react_estree::Program;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new(&buf);
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
