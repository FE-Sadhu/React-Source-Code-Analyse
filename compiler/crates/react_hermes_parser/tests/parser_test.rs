/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::env;

use insta::{assert_snapshot, glob};
use react_estree::SourceType;
use react_hermes_parser::parse;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let mut ast = parse(&input, path.to_str().unwrap()).unwrap();
        // TODO: hack to prevent changing lots of fixtures all at once
        ast.source_type = SourceType::Script;
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
