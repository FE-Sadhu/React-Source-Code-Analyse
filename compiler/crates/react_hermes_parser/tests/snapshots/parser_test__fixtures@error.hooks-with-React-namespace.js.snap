---
source: crates/react_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/react_hermes_parser/tests/fixtures/error.hooks-with-React-namespace.js
---
Input:
function Foo() {
  const [x, setX] = React.useState(1);
  return x;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Foo",
        "typeAnnotation": null,
        "loc": null,
        "range": [
          9,
          12
        ]
      },
      "params": [],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        26,
                        27
                      ]
                    },
                    {
                      "type": "Identifier",
                      "name": "setX",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        29,
                        33
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    25,
                    34
                  ]
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "React",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        37,
                        42
                      ]
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "useState",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        43,
                        51
                      ]
                    },
                    "computed": false,
                    "loc": null,
                    "range": [
                      37,
                      51
                    ]
                  },
                  "arguments": [
                    {
                      "type": "NumericLiteral",
                      "value": 1.0,
                      "loc": null,
                      "range": [
                        52,
                        53
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    37,
                    54
                  ]
                },
                "loc": null,
                "range": [
                  25,
                  54
                ]
              }
            ],
            "loc": null,
            "range": [
              19,
              55
            ]
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "Identifier",
              "name": "x",
              "typeAnnotation": null,
              "loc": null,
              "range": [
                65,
                66
              ]
            },
            "loc": null,
            "range": [
              58,
              67
            ]
          }
        ],
        "loc": null,
        "range": [
          15,
          69
        ]
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": [
        0,
        69
      ],
      "loc": null,
      "range": [
        0,
        69
      ]
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": [
    0,
    69
  ]
}
